File: typescript/Services/FileLoader.ts
import { readFile } from "node:fs/promises";
import { join } from "node:path";

import { Grid } from "../Models/Grid.ts";
import { RunType, SolutionType } from "../Models/SolutionModels.ts";

type Matrix<T> = T[][];

interface IFileLoader {
  loadRaw(day: number, solutionType: SolutionType, runType: RunType): Promise<string>;
  loadLines<T>(day: number, solutionType: SolutionType, runType: RunType): Promise<T[]>;
  loadGrid<T>(day: number, solutionType: SolutionType, runType: RunType): Promise<Grid<T>>;
  loadItems<T>(day: number, solutionType: SolutionType, runType: RunType): Promise<Matrix<T>>;
}

class FileLoader implements IFileLoader {
  private readonly solutionsRoot: string;

  public constructor(solutionsRoot: string = join(process.cwd(), "Solutions")) {
    this.solutionsRoot = solutionsRoot;
  }

  public async loadRaw(day: number, solutionType: SolutionType, runType: RunType): Promise<string> {
    const filePath = this.resolveDataFilePath(day, solutionType, runType);
    return readFile(filePath, { encoding: "utf-8" });
  }

  public async loadLines<T>(day: number, solutionType: SolutionType, runType: RunType): Promise<T[]> {
    const raw = await this.loadRaw(day, solutionType, runType);
    return raw
      .split(/\r?\n/)
      .map((line) => line.trimEnd())
      .filter((line) => line.length > 0)
      .map((line) => this.changeType<T>(line));
  }

  public async loadGrid<T>(day: number, solutionType: SolutionType, runType: RunType): Promise<Grid<T>> {
    const lines = await this.loadLines<string>(day, solutionType, runType);
    const grid = lines
      .map((line) => Array.from(line))
      .map((chars) => chars.filter((char) => char.trim().length > 0))
      .map((chars) => chars.map((char) => this.changeType<T>(char)));
    return new Grid(grid);
  }

  public async loadItems<T>(day: number, solutionType: SolutionType, runType: RunType): Promise<Matrix<T>> {
    const lines = await this.loadLines<string>(day, solutionType, runType);
    return lines
      .map((line) => line.split(/[\s\t]+/).filter((value) => value.length > 0))
      .map((entries) => entries.map((entry) => this.changeType<T>(entry)));
  }

  private resolveDataFilePath(day: number, solutionType: SolutionType, runType: RunType): string {
    const dayFolder = `Day${day.toString().padStart(2, "0")}`;
    const fileName = this.resolveFileName(solutionType, runType);
    return join(this.solutionsRoot, dayFolder, "Data", fileName);
  }

  private resolveFileName(solutionType: SolutionType, runType: RunType): string {
    const partIndex = solutionType === SolutionType.First ? "1" : "2";
    const prefix = runType === RunType.Test ? "Test" : "Full";
    return `${prefix}${partIndex}.txt`;
  }

  private changeType<T>(value: string): T {
    const trimmed = value.trim();

    if (/^-?\d+$/.test(trimmed)) {
      const numeric = Number(trimmed);
      if (Number.isSafeInteger(numeric)) {
        return numeric as T;
      }

      return BigInt(trimmed) as T;
    }

    if (/^-?\d+\.\d+$/.test(trimmed)) {
      return Number(trimmed) as T;
    }

    const lowered = trimmed.toLowerCase();
    if (lowered === "true" || lowered === "false") {
      return (lowered === "true") as T;
    }

    return trimmed as T;
  }
}

export type { IFileLoader, Matrix };
export { FileLoader };

File: dotnet/Services/FileLoader.cs
using AdventOfCode.Models;

namespace AdventOfCode.Services;

public interface IFileLoader
{
    Task<T> LoadRaw<T>(int day, SolutionType solutionType, RunType runType);
    Task<List<T>> LoadLines<T>(int day, SolutionType solutionType, RunType runType);
    Task<Grid<T>> LoadGrid<T>(int day, SolutionType solutionType, RunType runType);
    Task<List<List<T>>> LoadItems<T>(int day, SolutionType solutionType, RunType runType);
}

public class FileLoader(string? solutionsRoot = null) : IFileLoader
{
    private readonly string _solutionsRoot = solutionsRoot ?? Path.Combine(Directory.GetCurrentDirectory(), "Solutions");

    public async Task<T> LoadRaw<T>(int day, SolutionType solutionType, RunType runType)
    {
        var filePath = ResolveDataFilePath(day, solutionType, runType);
        return ChangeType<T>(await File.ReadAllTextAsync(filePath));
    }

    public async Task<List<T>> LoadLines<T>(int day, SolutionType solutionType, RunType runType)
    {
        var raw = await LoadRaw<string>(day, solutionType, runType);
        return [.. raw.Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries).Select(ChangeType<T>)];
    }

    public async Task<Grid<T>> LoadGrid<T>(int day, SolutionType solutionType, RunType runType)
    {
        var lines = await LoadLines<string>(day, solutionType, runType);
        List<List<T>> gridRows = [.. lines
            .Select(x => x.Select(y => y.ToString())
            .Where(y => !string.IsNullOrWhiteSpace(y))
            .Select(y => ChangeType<T>(y)).ToList())
        ];
        return new Grid<T>(gridRows);
    }

    public async Task<List<List<T>>> LoadItems<T>(int day, SolutionType solutionType, RunType runType)
    {
        var lines = await LoadLines<string>(day, solutionType, runType);
        var lineItems = lines.Select(line => line.Split([' ', '\t'], StringSplitOptions.RemoveEmptyEntries));
        return [.. lineItems.Select(x => x.Select(y => ChangeType<T>(y)).ToList())];
    }

    private static T ChangeType<T>(object obj)
    {
        return (T)Convert.ChangeType(obj, typeof(T));
    }

    private static string BuildDayFolder(int day) => $"Day{day:D2}";

    private static string BuildFileName(SolutionType solutionType, RunType runType)
    {
        var partSuffix = solutionType == SolutionType.First ? "1" : "2";
        var prefix = runType == RunType.Test ? "Test" : "Full";
        return $"{prefix}{partSuffix}.txt";
    }

    private string ResolveDataFilePath(int day, SolutionType solutionType, RunType runType)
    {
        return Path.Combine(_solutionsRoot, BuildDayFolder(day), "Data", BuildFileName(solutionType, runType));
    }
}